File: net/JetBrains.SignatureVerifier/src/Serialization/TextualInfo.cs
using System;
using System.Collections.Generic;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Math;

namespace JetBrains.SignatureVerifier.Serialization;

public abstract class TextualInfo : IEncodableInfo
{
  public static TextualInfo GetInstance(Asn1Encodable value)
  {
    switch (value)
    {
      case DerUtcTime utcTime:
        return new Asn1UtcTimeInfo(utcTime.ToDateTime());
      case DerGeneralizedTime generalizedTime:
        return new Asn1GeneralizedTimeInfo(generalizedTime.ToDateTime());
      case DerUniversalString universalString:
        return new DerUniversalStringInfo(universalString.GetOctets());
      case DerOctetString octetString:
        return new DerOctetStringInfo(octetString.GetOctets());
      case DerBitString bitString:
        return new DerBitStringInfo(bitString.GetOctets());
      case DerPrintableString printableString:
        return new DerPrintableStringInfo(printableString.GetString());
      case DerUtf8String utf8String:
        return new DerUTF8StringInfo(utf8String.GetString());
      case DerIA5String ia5String:
        return new DerIA5StringInfo(ia5String.GetString());
      case DerBmpString bmpString:
        return new DerBmpStringInfo(bmpString.GetString());
      case DerVisibleString visibleString:
        return new DerVisibleStringInfo(visibleString.GetString());
      case DerNumericString numericString:
        return new DerNumericStringInfo(numericString.ToString());
      case DerGeneralString generalString:
        return new DerGeneralStringInfo(generalString.ToString());
      case DerObjectIdentifier objectId:
        return new Asn1ObjectIdentifierInfo(objectId.ToString());
      case DerBoolean boolean:
        return new BooleanInfo(boolean.IsTrue);
      case DerInteger integer:
        return new IntegerInfo(integer.ToString());
      case DerNull _:
        return new Asn1NullInfo("NULL");
      default:
        throw new Exception($"{value.GetType().Name} is not handled in the factory method.");
    }
  }

  protected abstract Asn1Encodable ToEncodable();

  public Asn1Encodable ToPrimitive() => ToEncodable().ToAsn1Object();
}

public abstract class DateTextualInfo : TextualInfo
{
  protected DateTime Content { get; set; }
}

public class Asn1UtcTimeInfo : DateTextualInfo
{
  public Asn1UtcTimeInfo(DateTime content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerUtcTime(Content);
}

public class Asn1GeneralizedTimeInfo : DateTextualInfo
{
  public Asn1GeneralizedTimeInfo(DateTime content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerGeneralizedTime(Content);
}

public abstract class HexTextualInfo : TextualInfo
{
  protected byte[] Content { get; set; }
}

public class DerUniversalStringInfo : HexTextualInfo
{
  public DerUniversalStringInfo(byte[] content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerUniversalString(Content);
}

public class DerOctetStringInfo : HexTextualInfo
{
  public DerOctetStringInfo(byte[] content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerOctetString(Content);
}

public class DerBitStringInfo : HexTextualInfo
{
  public DerBitStringInfo(byte[] content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerBitString(Content);
}

public abstract class StringTextualInfo : TextualInfo
{
  protected string Content { get; set; }
}

public class DerPrintableStringInfo : StringTextualInfo
{
  public DerPrintableStringInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerPrintableString(Content);
}

public class DerUTF8StringInfo : StringTextualInfo
{
  public DerUTF8StringInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerUtf8String(Content);
}

public class DerIA5StringInfo : StringTextualInfo
{
  public DerIA5StringInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerIA5String(Content);
}

public class DerBmpStringInfo : StringTextualInfo
{
  public DerBmpStringInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerBmpString(Content);
}

public class DerVisibleStringInfo : StringTextualInfo
{
  public DerVisibleStringInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerVisibleString(Content);
}

public class DerNumericStringInfo : StringTextualInfo
{
  public DerNumericStringInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerNumericString(Content);
}

public class DerGeneralStringInfo : StringTextualInfo
{
  public DerGeneralStringInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerGeneralString(Content);
}

public class Asn1ObjectIdentifierInfo : StringTextualInfo
{
  public Asn1ObjectIdentifierInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerObjectIdentifier(Content);
}

public class BooleanInfo : TextualInfo
{
  private bool Content { get; }
  public BooleanInfo(bool content) => this.Content = content;

  protected override Asn1Encodable ToEncodable() =>
    DerBoolean.GetInstance(Content);
}
public class IntegerInfo : StringTextualInfo
{
  public IntegerInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => new DerInteger(new BigInteger(Content));
}

public class Asn1NullInfo : StringTextualInfo
{
  public Asn1NullInfo(string content) => this.Content = content;
  protected override Asn1Encodable ToEncodable() => DerNull.Instance;
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/TextualInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.BigIntegerSerializer
import com.jetbrains.signatureverifier.serialization.ByteArraySerializer
import com.jetbrains.signatureverifier.serialization.DateSerializer
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.*
import java.math.BigInteger
import java.util.*

/**
 * Holds various Strings, Numeric and other formats, that can be converted to String and back
 */
@Serializable
sealed interface TextualInfo :
  EncodableInfo {

  companion object {
    fun getInstance(value: ASN1Encodable): TextualInfo =
      when (value) {
        is DERPrintableString -> DERPrintableStringInfo(value.toString())
        is DERUTF8String -> DERUTF8StringInfo(value.toString())
        is DERIA5String -> DERIA5StringInfo(value.toString())
        is DERBMPString -> DERBMPStringInfo(value.toString())
        is DERVisibleString -> DERVisibleStringInfo(value.toString())
        is DERUniversalString -> DERUniversalStringInfo(value.octets)
        is DERNumericString -> DERNumericStringInfo(value.toString())
        is DERGeneralString -> DERGeneralStringInfo(value.toString())
        is DEROctetString -> DEROctetStringInfo(value.octets)
        is ASN1ObjectIdentifier -> ASN1ObjectIdentifierInfo(value.toString())
        is DERBitString -> DERBitStringInfo(value.octets)
        is ASN1Boolean -> BooleanInfo(value.toString())
        is ASN1Integer -> IntegerInfo(value.toString())
        is ASN1UTCTime -> ASN1UTCTimeInfo(value.date)
        is ASN1GeneralizedTime -> ASN1GeneralizedTimeInfo(value.date)
        is ASN1Enumerated -> ASN1EnumeratedInfo(value.value)
        is ASN1Null -> ASN1NullInfo(value.toString())
        else -> throw IllegalArgumentException("This type of strings is not in list: ${value::class}")
      }

  }

  fun toEncodable(): ASN1Encodable

  override fun toPrimitive(): ASN1Primitive = toEncodable().toASN1Primitive()
}

@Serializable
sealed class DateTextualInfo : TextualInfo {

  @Serializable(DateSerializer::class)
  abstract val content: Date
}

@Serializable
data class ASN1UTCTimeInfo(
  override val content: @Serializable(DateSerializer::class) Date
) : DateTextualInfo() {
  override fun toEncodable(): ASN1Encodable = ASN1UTCTime(content)
}

@Serializable
data class ASN1GeneralizedTimeInfo(
  override val content: @Serializable(DateSerializer::class) Date
) :
  DateTextualInfo() {
  override fun toEncodable(): ASN1Encodable = ASN1GeneralizedTime(content)
}

@Serializable
sealed class HexTextualInfo : TextualInfo {
  @Serializable(ByteArraySerializer::class)
  abstract val content: ByteArray
}

@Serializable
data class ASN1EnumeratedInfo(
  val content: @Serializable(BigIntegerSerializer::class) BigInteger
) : TextualInfo {
  override fun toEncodable(): ASN1Encodable = ASN1Enumerated(content)
}

@Serializable
data class DERUniversalStringInfo(
  @Serializable(ByteArraySerializer::class)
  override val content: ByteArray
) : HexTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERUniversalString(content)
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as DERUniversalStringInfo

    return content.contentEquals(other.content)
  }

  override fun hashCode(): Int {
    return content.contentHashCode()
  }
}

@Serializable
data class DEROctetStringInfo(
  @Serializable(ByteArraySerializer::class)
  override val content: ByteArray
) : HexTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DEROctetString(content)
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as DEROctetStringInfo

    return content.contentEquals(other.content)
  }

  override fun hashCode(): Int {
    return content.contentHashCode()
  }
}

@Serializable
data class DERBitStringInfo(
  @Serializable(ByteArraySerializer::class)
  override val content: ByteArray
) : HexTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERBitString(content)
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as DERBitStringInfo

    return content.contentEquals(other.content)
  }

  override fun hashCode(): Int {
    return content.contentHashCode()
  }
}

@Serializable
sealed class StringTextualInfo : TextualInfo {
  abstract val content: String
}

@Serializable
data class DERPrintableStringInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERPrintableString(content)
}

@Serializable
data class DERUTF8StringInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERUTF8String(content)
}

@Serializable
data class DERIA5StringInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERIA5String(content)
}

@Serializable
data class DERBMPStringInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERBMPString(content)
}

@Serializable
data class DERVisibleStringInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERVisibleString(content)
}

@Serializable
data class DERNumericStringInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERNumericString(content)
}

@Serializable
data class DERGeneralStringInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERGeneralString(content)
}

@Serializable
data class ASN1ObjectIdentifierInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = ASN1ObjectIdentifier(content)
}

@Serializable
data class BooleanInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = ASN1Boolean.getInstance(if (content == "TRUE") 1 else 0)
}

@Serializable
data class IntegerInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = ASN1Integer(content.toBigInteger())
}

@Serializable
data class ASN1NullInfo(override val content: String) : StringTextualInfo() {
  override fun toEncodable(): ASN1Encodable = DERNull.INSTANCE
}
